ui-charts: display a GOV in a GET in Storybook

jacomyal opened the pull request:
This PR addresses issue [#945](https://github.com/osrd-project/osrd-confidential/issues/945). It is huge, and has been designed to be reviewed **each commit one at a time**.

It was originally submitted [here](https://github.com/OpenRailAssociation/osrd-ui/pull/1050), on the OSRD-UI former repository.

The core parts are:
- Improving `Manchette` and `SpaceTimeChart` to ease the track-occupancy diagram integration
- Creating a new `TrackOccupancyStandalone` component, that properly uses `useManchetteWithSpaceTimeChart` to get a track-occupancy diagram, with the `SpaceTimeChart` interactions (panning and zooming)
- Adding a new story to showcase how to display a `TrackOccupancyCanvas` and a `TrackOccupancyCanvas` within a split section, using `useManchetteWithSpaceTimeChart`
- Re-implementing clicking trains in the `TrackOccupancyCanvas` layer, using picking

jacomyal committed:
ui-charts: improves splitting manchette and STC

Details:
- Adds splitting waypoints in the operational points given to the STC
  from useManchetteWithSpaceTimeChart, and handles not rendering them in
  the STC code directly
- Adds flatSteps in the SpaceTimeChartContext, as a Set<number>
- Steps rendering path segments on pauses on flat steps, since we now
  assume they are rendered directly in the split section layer itself

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal committed:
ui-charts: makes space-time chart more resilient

This commit allows the SpaceTimeChart to work with space scales that
just contain one single flat step. The idea is to allow using the
SpaceTimeChart, to only render a TrackOccupancyDiagram, while still
having the other interesting interactions (zooming and panning,
basically).

Details:
- Adds new hideTimeCaptions param
- Improves utils to allow the SpaceTimeChart to work with only one
  waypoint, and a flat step on it

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal committed:
ui-charts: makes manchette more resilient

This commit aims at making Manchette and useManchetteWithSpaceTimeChart
work with a single flat step, on a single waypoint. This will allow them
to be used to render a TrackOccupancyDiagram, with all the interesting
interactions (zooming and panning).

Details:
- Adds various options to useManchetteWithSpaceTimeChart, to allow
  disabling zooming, panning or time captions
- Allows overriding vertical padding in useManchetteWithSpaceTimeChart,
  to exactly fit one single flat step, to the pixel
- Improves helpers (and their tests) to work in the edge case where
  there is just one waypoint, and a flat step on it

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal committed:
ui-charts: adds new TrackOccupancyStandalone

This commit basically replaces the previous
trackOccupancyDiagram/rendering story, with a new component,
TrackOccupancyStandalone, that actually calls
useManchetteWithSpaceTimeChart instead of mimicking it. The main benefit
is to get proper interactions, such as panning on the time axis.

Also, this commit is very impactful, because many things were done quite
customly for track occupancy drawing, when the SpaceTimeChart APIs
provide many tools to help rendering additional layers on top of them.

One final note, before getting into the detail: This commit breaks
selecting a train with the mouse on the TrackOccupancyDiagram, because
this was done in the rendering loop, rather than using picking (as it
should).

Details:
- Replaces all time and space to pixels conversions from track occupancy
  layers, with code that uses proper tools from SpaceTimeChartContext
- Removes mouse collision detection (it was done in the rendering
  process, which meant that rendering was depending on the
  mousePositions, causing many excess renderings)
- Updates OccupancyZone: id becomes trainId, and arrivalTime and
  departureTime become numbers instead of dates
- Adds new TrackOccupancyStandalone component
- Updates rendering story to use new component

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal committed:
ui-charts: fixes some manchette stories

Some code from useManchetteWithSpaceTimeChart stories was out of date,
this commit fixes those occurances.

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal committed:
ui-charts: normalizes millimeters in stories

There were some weird issues due to the fact that data for the
manchette and useManchetteWithSpaceTimeChart were in millimeters, and
data for the SpaceTimeChart were in meters.

This commit migrates everything SpaceTimeChart-related to millimeters.

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal committed:
ui-charts: drafts TOD within STC story

This commit addresses ticket:
https://github.com/osrd-project/osrd-confidential/issues/945

Details:
- Adds possibility to add some top padding to TrackOccupancyCanvas
- Adds possibility to display a close button in the top right of the
  TrackOccupancyCanvas
- Allows giving children to TrackOccupancyManchette
- Adds Waypoint (as WaypointComponent) to exports from ui-charts
- Creates new trac-occupancy story, that shows how to display
  TrackOccupancy diagrams within split sections of the Manchette and the
  SpaceTimeChart

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal committed:
ui-charts: restores clicking trains in the TOD

Details:
- Implements back clicks detection in the TrackOccupancy diagram, using
  the picking framework from the SpaceTimeDiagram
- Adds an example showing how it works in the
  trackOccupancyDiagram/rendering story

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal committed:
ui-charts: adds top and bottom borders to TOD

Details:
- Adds top and bottom borders to TrackOccupancyCanvas
- Adds top and bottom borders to TrackOccupancyManchette
- Removes these borders when using TrackOccupancyStandalone

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

clarani commented on front/ui/ui-charts/src/spaceTimeChart/components/PathLayer.tsx:
```diff
@@ -194,9 +214,11 @@ export const PathLayer = ({
-            // Detect flat steps, and draw two graduations if any (one on each side of the step):
-            getSpacePixels(getSpacePixel, position).forEach((rawPixel) => {
-              const spacePixel = getCrispLineCoordinate(rawPixel, ctx.lineWidth);
+            // Only draw the stop when there is no flat step
+            // (i.e. when there's only one space pixel):
+            const rawPixels = getSpacePixels(getSpacePixel, position);
```
I didn't understand why, if it was a flat step, there would be more or less than one pixel :thinking:

jacomyal replied:
That's what `getSpacePixels` does:
https://github.com/OpenRailAssociation/osrd/blob/fa5b8505653b59e97a68f9b5bd4aa7212a5b9f22/front/ui/ui-charts/src/spaceTimeChart/utils/paths.ts#L33-L37

Basically, if there is a flat step on the `position` value, then `getSpacePixels` returns `[topPixelValue, bottomPixelValue]`. Else, it returns `[pixelValue]`.

clarani replied:
Oh thanks, it's clear :white_check_mark:

clarani commented on front/ui/ui-charts/src/spaceTimeChart/components/PathLayer.tsx:
```diff
@@ -351,11 +373,13 @@ export const PathLayer = ({
-          totalLength += Math.sqrt(Math.pow(prevX - x, 2) + Math.pow(prevY - y, 2));
-        }
+      lines.forEach((points) => {
+        points.forEach(({ x, y }, i, a) => {
```
```suggestion
      lines.forEach((line) => {
        line.forEach(({ x, y }, i, points) => {
```

clarani commented on front/ui/ui-charts/src/spaceTimeChart/utils/scales.ts:
```diff
@@ -188,11 +188,16 @@ export function getSpaceToPixel(
     );
+    // Rare case where coefficient is 0:
+    // (occurs when there is just a flat step, for instance)
+    if (!coefficient) return pixelOffset + (fromEnd ? pixelTo : pixelFrom);
```
nit
```suggestion
    if (coefficient === 0) return pixelOffset + (fromEnd ? pixelTo : pixelFrom);
```

jacomyal replied:
I'm applying the diff, but just, out of curiosity, why do you prefer `coefficient === 0`?

clarani replied:
It's just because it's more explicit. You specify 2 lines above that you test that coefficient is 0, so it seems more natural to see `coefficient === 0`

clarani commented on front/ui/ui-charts/src/manchette/hooks/useManchetteWithSpaceTimeChart.tsx:
```diff
@@ -643,18 +675,20 @@ const useManchetteWithSpaceTimeChart = ({
+            if (enableSpacePan) {
+              let newYOffset = initialOffset.y - diff.y;
+              newYOffset = Math.max(newYOffset, 0);
+              if (manchette)
+                newYOffset = Math.min(newYOffset, manchette.scrollHeight - manchette.offsetHeight);
+              newState.yOffset = newYOffset;
+              if (manchette) manchette.scrollTop = newYOffset;
```
nit: only check manchette once
```suggestion
              if (manchette){
                newYOffset = Math.min(newYOffset, manchette.scrollHeight - manchette.offsetHeight);
                manchette.scrollTop = newYOffset;
              }
              newState.yOffset = newYOffset;
```

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/TrackOccupancyStandalone.tsx:
Nice component ! I'm just wondering, isn't it a bit weird to use the SpaceTimeChart in this component ?

Maybe for later, we should create a special hook to handle interactions for the TrackOccupancyDiagram, and make it independent from the SpaceTimeChart

SharglutDev replied:
Or maybe (just some random ideas) the hook could be renamed `useManchetteWithChart` and `SpaceTimeChart` and `TrackOccupancyStandalone` could both call a kind of generic `CommonChart` / `BaseChart` component so it would be less "weird" ?

jacomyal replied:
After our discussion from last week, I suggest we keep things as they are for now, and we can start plan some refactoring for later.

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/TrackOccupancyStandalone.tsx:
```diff
@@ -0,0 +1,105 @@
+  const spaceTimeChartRef = useRef<HTMLDivElement>(null);
+  const defaultTimeOrigin = useMemo(() => {
+    const minTime = Math.min(...(occupancyZones.map((zone) => zone.arrivalTime) || Date.now()));
+    // Take first round hour before minTime:
+    return Math.floor(minTime / HOUR) * HOUR;
```
Nice :+1:

clarani commented on front/ui/storybook/stories/ui-charts/trackOccupancyDiagram/rendering.stories.tsx:
```diff
@@ -1,286 +1,38 @@
-import { KebabHorizontal } from '../../../../ui-icons/src/index';
-import { OPERATIONAL_POINTS } from '../spaceTimeChart/helpers/paths';
+import OCCUPANCY_ZONES from './assets/occupancyZones';
+import { TRACKS } from '../manchetteWithSpaceTimeChart/assets/trackOccupancyData';
```
This change breaks the build at this commit
```suggestion
import TRACKS from './assets/tracks';
```

clarani commented on front/ui/storybook/stories/ui-charts/trackOccupancyDiagram/rendering.stories.tsx:
```diff
@@ -1,286 +1,38 @@
-import React, { useEffect, useMemo, useState } from 'react';
+import React, { useEffect, useState } from 'react';
+
+import { TrackOccupancyStandalone } from '@osrd-project/ui-charts';
```
Extra empty line below this import (you can rebase to have the eslint configured correctly)

clarani commented on front/ui/storybook/stories/ui-charts/trackOccupancyDiagram/rendering.stories.tsx:
```diff
@@ -1,286 +1,38 @@
+  trainId: number;
+  autoHeight?: boolean;
+}) => {
+  const [selectedTrainId, setSelectedTrainId] = useState<string | undefined>(undefined);
```
nit
```suggestion
  const [selectedTrainId, setSelectedTrainId] = useState<string>();
```

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/helpers/drawElements/drawTracks.ts:
```diff
@@ -1,90 +1,48 @@
+  } = stcContext;
+  const yStart = getSpacePixel(position);
+  const yEnd = getSpacePixel(position, true);
+  const height = yEnd - yStart;
```
Why not get height from the context ? :thinking:

jacomyal replied:
Because this `height` is the height of the flat step, while the `height` in the context is the full SpaceTimeChart's height in pixels.

I'll rename the variable as `flatStepHeight`.

clarani replied:
Thanks, it's clearer

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/helpers/drawElements/drawOccupancyZones.ts:
```diff
@@ -171,77 +138,86 @@ const drawOccupationZone = ({
-    let lastDepartureTimePixel = primaryDepartureTimePixel;
+    let primaryArrivalTime = 0;
+    let primaryDepartureTime = 0;
+    let lastDepartureTime = primaryDepartureTime;
```
```suggestion
    let lastDepartureTime = 0;
```

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/helpers/drawElements/drawOccupancyZones.ts:
```diff
@@ -171,77 +138,86 @@ const drawOccupationZone = ({
+    });
+  }
+
+  // Draw trains:
```
Hum what about draw train links ? draw links to trains ?

jacomyal replied:
Yes indeed, this comment is quite bad ^^ I'll update it, thanks for the feedback :+1:

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/helpers/drawElements/drawOccupancyZones.ts:
```diff
@@ -97,71 +90,45 @@ const drawRemainingTrainsBox = ({ ctx, remainingTrainsNb, xPosition }: DrawRemai
+  {
+    zone,
+    position,
+    yZone,
```
Hum position and yZone are not very clear.
Actually, `position` corresponds to the position on path and `yZone` corresponds to the y position vertically centered on the current track ?

jacomyal replied:
I can't find this code anymore, I think it's been removed in some of the next commits in the PR.

clarani commented on front/ui/storybook/stories/ui-charts/spaceTimeChart/helpers/consts.ts:
```diff
@@ -20,4 +20,4 @@ export const MINUTE = 60 * SECOND;
 export const HOUR = 60 * MINUTE;
 
 // Same for distances in meters:
-export const KILOMETER = 1000;
```
You need to change the comment (it's distances in mm)

SharglutDev commented on front/ui/storybook/stories/ui-charts/manchetteWithSpaceTimeChart/track-occupancy.stories.tsx:
Played a bit with the TOD (track occupancy diagram) and there is a weird behavior with horizontal zoom (same issue on the split point stories) : 


https://github.com/user-attachments/assets/1e06fdbc-124f-47d5-9935-8a956df6be99

Maybe it could be nice to have the slider on this story ?

Also do we have a spec of how the rectangular zoom works on the TOD ?

jacomyal replied:
I can't reproduce the issue on the video on my computer (I tested Chromium and Firefox). About the rectangle selection, I suggest we keep going forward on that PR (to avoid polluting the stories), and we test and fix that issue once we work on the [OSRD integration](https://github.com/osrd-project/osrd-confidential/issues/866), directly in OSRD.

SharglutDev replied:
To reproduce, I click on the GOV, I hold "shift" and just scroll bottom or up

SharglutDev replied:
Agreed for the horizontal zoom / rectangle zoom

SharglutDev replied:
Also on the standalone GOV, I can also hold "shift" and it's scrolling in the manchette part while the GOV part stays static

jacomyal replied:
@clarani I can't reproduce this behaviour (i.e. it behaves as expected, zooming in the GET without scrolling in the manchette) on my Linux laptop. I tested on Chrome and Firefox, with my mouse and my laptop. @SharglutDev tells me it occurs on his mac.

Can you please test to see if you can reproduce it, so that I get more hints on what's going on?

SharglutDev replied:
I pinged some mac users as well to see if they have the same issue

Akctarus replied:
On Mac I can reproduce it on Safari and Firefox

Yohh replied:
same thing on mac with safari, brave and chrome, note that it only happens with the trackpad, with the mouse wheel I can't even zoom

Yohh replied:
another thing: scrolling after having zoom fixes the offset

clarani replied:
I can't reproduce (Ubuntu, Firefox)

jacomyal replied:
OK, thanks all for your feedbacks :pray: 

I actually can't find any matching `e.preventDefault` or `e.stopPropagation`. Basically, the Mac behaviours seems to make more sense than the other behaviour... And even weirder: It actually looks like, on my computer, Shift+Wheel never scrolls in web pages, even like on this GitHub page, right now... That would explain. I'll try to find how to disable that behaviour so that I can fix the issue.

Thanks again!

jacomyal replied:
@SharglutDev I replaced locally `Shift` with `Alt`, which allowed me to reproduce the issue. I supposedly fixed it, with a `e.preventDefault()` at the right place. I just pushed a fix.

jacomyal replied:
I'm closing this conversation so that I can merge the PR. I've been told the issue is still occurring, but I still can't reproduce it locally, and it should probably not be linked to that PR, and should have been occurring before, without the TrackOccupancyDiagram.

clarani commented on front/ui/storybook/stories/ui-charts/spaceTimeChart/helpers/paths.ts:
```diff
@@ -64,37 +65,37 @@ export const OPERATIONAL_POINTS: OperationalPoint[] = [
     id: 'city-a',
     label: 'Point A',
-    position: 0 * KM,
+    position: 0 * KILOMETER,
```
Hum I'm hesitating. Maybe we should use the helper `positionMmToKm`, this way the reader will immidiately know that it needs to provide positions in mm.
What do you think about this?

jacomyal replied:
I agree that this version doesn't make it clear that it's expecting millimeters. But I don't get your suggestion: `positionMmToKm` does indeed take millimeters as input, but it returns the position in kilometers, which is not what we want here.

What if I just add another comment to the `OperationalPoint` type declaration, to specify that it's expecting millimeters?

clarani replied:
Ah sorry, what I meant was to add a utils function kmToMm

clarani commented on front/ui/storybook/stories/ui-charts/spaceTimeChart/horizontal-zoom.stories.tsx:
```diff
@@ -73,7 +73,7 @@ const SpaceTimeHorizontalZoomWrapper = ({
       from: 0,
       to: 75000,
-      coefficient: 300,
+      coefficient: 300000,
```
Why change the coefficient and not the `to` position ?

clarani commented on front/ui/storybook/stories/ui-charts/spaceTimeChart/rectangle-zoom.stories.tsx:
```diff
@@ -30,9 +31,9 @@ const MAX_ZOOM = 100;
-const MIN_ZOOM_METER_PER_PX = 10000;
-const MAX_ZOOM_METER_PER_PX = 10;
-const DEFAULT_ZOOM_METER_PER_PX = 300;
+const MIN_SPACE_ZOOM = 10 * KILOMETER;
+const MAX_SPACE_ZOOM = 0.01 * KILOMETER;
+const DEFAULT_SPACE_ZOOM = 0.3 * KILOMETER;
```
I'm ok with the renaming, but can you add some comments to specify that the distance given is max/min per px ?

SharglutDev commented on front/ui/ui-charts/src/manchette/hooks/useManchetteWithSpaceTimeChart.tsx:
nit : we shoudl'nt use `GET` in commits, it's a french acronym :)

clarani commented on front/ui/storybook/stories/ui-charts/spaceTimeChart/helpers/components.tsx:
```diff
@@ -90,12 +90,12 @@ const DataLabel = ({
         <>
           <div>Time difference: {formatTimeLength(new Date(data.time))}</div>
-          <div>Distance to mark: {round(data.position).toLocaleString()} m</div>
+          <div>Distance to mark: {round(data.position / KILOMETER).toLocaleString()} km</div>
```
Can you use the helper `positionMmToKm` instead of dividing by KILOMETER here ?

clarani commented on front/ui/storybook/stories/ui-charts/spaceTimeChart/split.stories.tsx:
Do you need to change the `to` position here too ?

clarani commented on front/ui/storybook/stories/ui-charts/manchetteWithSpaceTimeChart/track-occupancy.stories.tsx:
```diff
@@ -0,0 +1,154 @@
+ */
+const TrackOccupancyDiagramWithinSpaceTimeChartWrapper = ({ height = 561 }: { height: number }) => {
+  // TODO: Restore trains selection from GOV
+  const [selectedTrain, _setSelectedTrain] = useState<undefined | string>(undefined);
```
nit:
```suggestion
  const [selectedTrain, _setSelectedTrain] = useState<string>();
```

clarani commented on front/ui/storybook/stories/ui-charts/spaceTimeChart/helpers/paths.ts:
```diff
@@ -240,3 +246,68 @@ export const PATHS: PathDisplay[] = [
+            : 'down',
+        arrivalTime: prevTime,
+        departureDirection: !next ? undefined : next.position < position ? 'up' : 'down',
+        departureTime: time,
```
Hum so you're going to add several items in the result array for 1 occupation zone right ? Why not directly look for the last item in the array with the waypointPosition ?

jacomyal replied:
A single path can in theory stop at different times on the same waypoint (such as a train going back and forth), and the case does occur in the fake sample data. You can check the green selected train on the following image, for instance:

![image](https://github.com/user-attachments/assets/0295e7ab-9968-47fd-8712-ac7bfa9a2c49)

clarani replied:
I was concerned that you add several zones for consecutive points, which would belong to the same stop. But it does not seem to break anything with your data test, so I think it's ok

clarani commented on front/ui/storybook/stories/ui-charts/spaceTimeChart/helpers/paths.ts:
```diff
@@ -240,3 +246,68 @@ export const PATHS: PathDisplay[] = [
   ),
 ];
+
+export function getOccupancyZonesFromPath<T extends object>(
```
```suggestion
export function getOccupancyZonesFromPathAtGivenWaypoint<T extends object>(
```

SharglutDev commented on front/ui/ui-charts/src/spaceTimeChart/components/PathLayer.tsx:
```diff
@@ -106,28 +106,39 @@ export const PathLayer = ({
+      for (let i = 0; i < points.length; i++) {
+        const { position, time } = points[i];
+
         if (!i) {
```
nit : Could we put that instead ?
```suggestion
        if (i === 0) {
```

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/layers/OccupancyZonesLayer.tsx:
```diff
@@ -17,19 +54,168 @@ const OccupancyZonesLayer = ({
+          instructions.push({
+            type: 'remainingTrains',
+            amount: remainingTrainsNb,
+            time: (zone.arrivalTime + zone.departureTime) / 2,
```
You need to compute the centered time of the displayed trains in the pack to display the remainingTrains label correctly

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/layers/OccupancyZonesLayer.tsx:
```diff
@@ -17,19 +54,168 @@ const OccupancyZonesLayer = ({
   topPadding: number;
   selectedTrainId?: string;
 }) => {
+  const zonesToDraw = useMemo(() => {
```
Why not instructionsToDraw ?

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/layers/OccupancyZonesLayer.tsx:
```diff
@@ -17,19 +54,168 @@ const OccupancyZonesLayer = ({
+      zonesToDraw.forEach((instruction) => {
+        if (instruction.type === 'occupancyZone') {
+          const x = getTimePixel(instruction.zone.arrivalTime);
+          const y = instruction.offsetY + getSpacePixel(position);
```
getSpacePixel(position) corresponds to the global y offset right ? Maybe you can declare it outside the forEach loop

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/layers/OccupancyZonesLayer.tsx:
```diff
@@ -1,8 +1,45 @@
+} from '../helpers/drawElements/drawOccupancyZones';
+import type { OccupancyZone, OccupancyZonePickingElement, Track } from '../types';
+
+interface BaseRenderingInstruction {
+  type: string;
+  offsetY: number;
+}
+interface OccupancyZoneRenderingInstruction extends BaseRenderingInstruction {
+  type: 'occupancyZone';
+  zone: OccupancyZone;
+  isSelected: boolean;
+}
+interface RemainingTrainsRenderingInstruction extends BaseRenderingInstruction {
+  type: 'remainingTrains';
+  amount: number;
+  time: number;
+}
+type RenderingInstruction = OccupancyZoneRenderingInstruction | RemainingTrainsRenderingInstruction;
```
Hum I find this clearer, what do you think ?
Option 1
```
type RenderingInstruction = { offsetY: number } & (
  | {
    type: 'occupancyZone';
    zone: OccupancyZone;
    isSelected: boolean;
  }
  | {
    type: 'remainingTrains';
    amount: number;
    time: number;
  }
);
```

Option 2
```
type RenderingInstruction = (
  | {
    type: 'occupancyZone';
    zone: OccupancyZone;
    isSelected: boolean;
    offsetY: number
  }
  | {
    type: 'remainingTrains';
    amount: number;
    time: number;
    offsetY: number
  }
);
```

clarani commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/helpers/drawElements/drawOccupancyZones.ts:
```diff
@@ -114,12 +107,13 @@ const drawOccupationZone = (
 
   const { getTimePixel, getSpacePixel } = stcContext;
-  const yStart = getSpacePixel(position);
+  const yStart = getCrispLineCoordinate(getSpacePixel(position), BACKGROUND_HEIGHT);
```
What is this yStart ?

jacomyal replied:
`yStart` is the top y coordinate of the flat step that contains the occupations zones layer.

SharglutDev commented on front/ui/ui-charts/src/trackOccupancyDiagram/components/TrackOccupancyStandalone.tsx:
```diff
@@ -0,0 +1,125 @@
+    defaultTimeOrigin,
+    verticalPadding: 0,
+    options: {
+      displayTimeCaptions: false,
+      enableTimePan: true,
+      enableSpacePan: true,
+      enableTimeZoom: false,
```
Don't we want a way to control these options with props when using the standalone TOD ?

> [!NOTE]
> On the standalone [mockup](https://www.sketch.com/s/50744654-4631-4fae-8e7c-7eb22dfadc1a/f/5D330E6B-6EE9-42DB-A845-1777C3B4860F), it looks like the time graduations are displayed (it's also not that simple, the current time gradutions position displayed on the space time chart are not matching the one from the mockup)

jacomyal committed:
ui-charts: integrates various minor changes

This commit addresses various discussions opened by reviewers on PR
https://github.com/OpenRailAssociation/osrd/pull/11636

Details:
- Adds train selection in track-occupancy.stories.tsx
- Fixes "+XXX trains" label placement in OccupancyZonesLayer
- Adds various comments and renames some variables and functions to
  improve code readability

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal committed:
ui-charts: draws times in TrackOccupancyStandalone

This commit adds times graduations in the bottom of the
TrackOccupancyStandalone component.

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

clarani commented on front/ui/storybook/stories/ui-charts/spaceTimeChart/helpers/paths.ts:
```diff
@@ -1,9 +1,9 @@
+  type OperationalPoint,
+  type PathData,
+  type PathLevel,
 } from '@osrd-project/ui-charts';
```
nit
```suggestion
import type {
  DataPoint,
  OccupancyZone,
  OperationalPoint,
  PathData,
  PathLevel,
} from '@osrd-project/ui-charts';
```

clarani reviewed (commented):
The more trains label is not centered anymore
![image](https://github.com/user-attachments/assets/b2f60689-31fa-4e8b-8555-ac81763a3235)

jacomyal committed:
ui-charts: fixes shift+wheel on SpaceTimeChart

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal committed:
ui-charts: integrates various minor changes

This commit addresses various discussions opened by reviewers on PR
https://github.com/OpenRailAssociation/osrd/pull/11636

Details:
- Fixes "+XXX trains" label placement in OccupancyZonesLayer, so that
  it's properly centered on displayed zones of the pack
- Replaces `XXX * KILOMETER` with `positionKmToMm(XXX)` for clarity

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal commented:
> The more trains label is not centered anymore

Oh, ok, I misread your comment on the "+X trains" label coordinates. I just pushed a fix.

SharglutDev reviewed (approved):
Lgtm and tested (tested on ubuntu), great job :white_check_mark: 
Lets see if it works for the mac users :)

clarani reviewed (approved):
LGTM :white_check_mark:
